package com.playtube.protube.video.music.local.subscription;

import static org.schabi.newpipe.extractor.subscription.SubscriptionExtractor.ContentSource.CHANNEL_URL;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.text.util.Linkify;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.appcompat.app.ActionBar;
import androidx.core.text.util.LinkifyCompat;

import com.evernote.android.state.State;

import com.playtube.protube.video.music.BaseFragment;
import com.playtube.protube.video.music.R;
import com.playtube.protube.video.music.ads.AdUtils;
import com.playtube.protube.video.music.error.ErrorInfo;
import com.playtube.protube.video.music.error.ErrorUtil;
import com.playtube.protube.video.music.error.UserAction;
import org.schabi.newpipe.extractor.NewPipe;
import org.schabi.newpipe.extractor.exceptions.ExtractionException;
import org.schabi.newpipe.extractor.subscription.SubscriptionExtractor;
import com.playtube.protube.video.music.local.subscription.workers.SubscriptionImportInput;
import com.playtube.protube.video.music.streams.io.NoFileManagerSafeGuard;
import com.playtube.protube.video.music.streams.io.StoredFileHelper;
import com.playtube.protube.video.music.util.Constants;
import com.playtube.protube.video.music.util.PermissionHelper;
import com.playtube.protube.video.music.util.ServiceHelper;

import java.util.Collections;
import java.util.List;

public class SubscriptionsImportFragment extends BaseFragment {
    private static final String KEY_SERVICE_ID = "service_id";
    @State
    int currentServiceId = Constants.NO_SERVICE_ID;

    private List<SubscriptionExtractor.ContentSource> supportedSources;
    private String relatedUrl;

    @StringRes
    private int instructionsString;

    /*//////////////////////////////////////////////////////////////////////////
    // Views
    //////////////////////////////////////////////////////////////////////////*/

    private TextView infoTextView;
    private EditText inputText;
    private Button inputButton;

    private final ActivityResultLauncher<Intent> requestImportFileLauncher =
            registerForActivityResult(new StartActivityForResult(), this::requestImportFileResult);

    public static SubscriptionsImportFragment getInstance(final int serviceId) {
        final SubscriptionsImportFragment instance = new SubscriptionsImportFragment();
        final Bundle args = new Bundle();
        args.putInt(KEY_SERVICE_ID, serviceId);
        instance.setArguments(args);
        return instance;
    }

    private void setInitialData(final int serviceId) {
        this.currentServiceId = serviceId;
    }

    ///////////////////////////////////////////////////////////////////////////
    // Fragment LifeCycle
    ///////////////////////////////////////////////////////////////////////////

    @Override
    public void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (currentServiceId == Constants.NO_SERVICE_ID && getArguments() != null) {
            setInitialData(getArguments().getInt(KEY_SERVICE_ID, Constants.NO_SERVICE_ID));
        }

        setupServiceVariables();
        if (supportedSources.isEmpty() && currentServiceId != Constants.NO_SERVICE_ID) {
            ErrorUtil.showSnackbar(activity,
                    new ErrorInfo(new String[]{}, UserAction.SUBSCRIPTION_IMPORT_EXPORT,
                            "Service does not support importing subscriptions",
                            currentServiceId,
                            R.string.general_error));
            activity.finish();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        setTitle(getString(R.string.import_title));
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater,
                             @Nullable final ViewGroup container,
                             final Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_import, container, false);
    }

    /*/////////////////////////////////////////////////////////////////////////
    // Fragment Views
    /////////////////////////////////////////////////////////////////////////*/

    @Override
    protected void initViews(final View rootView, final Bundle savedInstanceState) {
        super.initViews(rootView, savedInstanceState);

        inputButton = rootView.findViewById(R.id.input_button);
        inputText = rootView.findViewById(R.id.input_text);

        infoTextView = rootView.findViewById(R.id.info_text_view);

        // TODO: Support services that can import from more than one source
        //  (show the option to the user)
        if (supportedSources.contains(CHANNEL_URL)) {
            inputButton.setText(R.string.import_title);
            inputText.setVisibility(View.VISIBLE);
            inputText.setHint(ServiceHelper.getImportInstructionsHint(currentServiceId));
        } else {
            inputButton.setText(R.string.import_file_title);
        }

        if (instructionsString != 0) {
            if (TextUtils.isEmpty(relatedUrl)) {
                setInfoText(getString(instructionsString));
            } else {
                setInfoText(getString(instructionsString, relatedUrl));
            }
        } else {
            setInfoText("");
        }

        final ActionBar supportActionBar = activity.getSupportActionBar();
        if (supportActionBar != null) {
            supportActionBar.setDisplayShowTitleEnabled(true);
            setTitle(getString(R.string.import_title));
        }

        final FrameLayout bannerAdContainer = rootView.findViewById(R.id.bannerAdContainer);
        AdUtils.loadGoogleBanner(requireActivity(), bannerAdContainer);
    }

    @Override
    protected void initListeners() {
        super.initListeners();
        inputButton.setOnClickListener(v -> onImportClicked());
    }

    private void onImportClicked() {
        if (!PermissionHelper.checkPostNotificationsPermission(requireActivity(),
                PermissionHelper.POST_NOTIFICATIONS_REQUEST_CODE)) {
            return;
        }

        if (inputText.getVisibility() == View.VISIBLE) {
            final String value = inputText.getText().toString();
            if (!value.isEmpty()) {
                onImportUrl(value);
            }
        } else {
            onImportFile();
        }
    }

    public void onImportUrl(final String value) {
        ImportConfirmationDialog.show(this,
                new SubscriptionImportInput.ChannelUrlMode(currentServiceId, value));
    }

    public void onImportFile() {
        NoFileManagerSafeGuard.launchSafe(
                requestImportFileLauncher,
                // leave */* mime type to support all services
                // with different mime types and file extensions
                StoredFileHelper.getPicker(activity, "*/*"),
                TAG,
                getContext()
        );
    }

    private void requestImportFileResult(final ActivityResult result) {
        final String data = result.getData() != null ? result.getData().getDataString() : null;
        if (result.getResultCode() == Activity.RESULT_OK && data != null) {
            ImportConfirmationDialog.show(this,
                    new SubscriptionImportInput.InputStreamMode(currentServiceId, data));
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // Subscriptions
    ///////////////////////////////////////////////////////////////////////////

    private void setupServiceVariables() {
        if (currentServiceId != Constants.NO_SERVICE_ID) {
            try {
                final SubscriptionExtractor extractor = NewPipe.getService(currentServiceId)
                        .getSubscriptionExtractor();
                supportedSources = extractor.getSupportedSources();
                relatedUrl = extractor.getRelatedUrl();
                instructionsString = ServiceHelper.getImportInstructions(currentServiceId);
                return;
            } catch (final ExtractionException ignored) {
                // Ignore and fall back to defaults below.
            }
        }

        supportedSources = Collections.emptyList();
        relatedUrl = null;
        instructionsString = 0;
    }

    private void setInfoText(final String infoString) {
        infoTextView.setText(infoString);
        LinkifyCompat.addLinks(infoTextView, Linkify.WEB_URLS);
    }
}

